### PR TITLE
crimson/os/seastore: use writev to avoid extra memory copy

### PR DESCRIPTION
**Pull request by xxhdx1985126**
Signed-off-by: Xuehan Xu <xxhdx1985126@gmail.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/crimson/os/seastore/journal.cc [hunk 1]
```diff
@@ -147,6 +147,13 @@ ceph::bufferlist Journal::encode_record(
     sizeof(checksum_t),
     reinterpret_cast<const char *>(&metadata_crc_le));
 
+  bufferptr bp(ceph::buffer::create_aligned(bl.length(), block_size));
+  bp.zero();
+  auto iter = bl.cbegin();
+  iter.copy(bl.length(), bp.c_str());
+  bl.clear();
+  bl.append(bp);
```
Comment (tchaikov) [inline src/crimson/os/seastore/journal.cc:155]: i think this is equivalent to `bl.rebuild()`, am i right?

File: src/crimson/os/seastore/segment_manager/block.cc [hunk 2]
```diff
@@ -291,6 +322,19 @@ Segment::write_ertr::future<> BlockSegment::write(
   return manager.segment_write({id, offset}, bl);
 }
 
+Segment::write_ertr::future<> BlockSegment::writev(
+  segment_off_t offset, ceph::bufferlist bl)
+{
+  if (offset < write_pointer || offset % manager.superblock.block_size != 0)
```
Comment (tchaikov) [inline src/crimson/os/seastore/segment_manager/block.cc:328]: i'd suggest add curly brackets around the statement even if it is just a single line, just for better readability.

File: src/crimson/os/seastore/journal.cc (hunk 1 above)
Comment (xxhdx1985126) [reply src/crimson/os/seastore/journal.cc:155]: That's because data sent to the underlying blk layer has to be aligned to the block size

Comment (xxhdx1985126) [reply src/crimson/os/seastore/journal.cc:155]: I mean each iovec has to be aligned to the block size

Comment (xxhdx1985126) [reply src/crimson/os/seastore/journal.cc:155]: Oh, sorry, I misunderstood you, will correct it.

**Comment by xxhdx1985126** [comment]
@tchaikov I've correct the code as suggested, please take a look, again, thanks:-)

File: src/crimson/os/seastore/segment_manager/block.cc [hunk 3]
```diff
@@ -46,6 +46,37 @@ static write_ertr::future<> do_write(
...
+static write_ertr::future<> do_writev(
+  seastar::file &device,
+  uint64_t offset,
+  bufferlist &bl)
+{
+  logger().debug(
+    "block: do_writev offset {} len {}",
+    offset,
+    bl.length());
+  std::vector<iovec> iov;
```
Comment (tchaikov) [inline src/crimson/os/seastore/segment_manager/block.cc:58]: might want to `reserve()`?

File: src/crimson/os/seastore/segment_manager/block.cc [hunk 4]
```diff
@@ -46,6 +46,37 @@ static write_ertr::future<> do_write(
...
+  for (auto& bp : bl.mut_buffers()) {
+    logger().debug("block: do_writev iov length: {}", bp.length());
+    iov.emplace_back(iovec{bp.c_str(), bp.length()});
+  }
+  return device.dma_write(
+    offset,
+    iov
```
Comment (tchaikov) [inline src/crimson/os/seastore/segment_manager/block.cc:65]:
```suggestion
    std::move(iov)
```

File: src/crimson/os/seastore/journal.cc [hunk 5]
```diff
@@ -147,6 +147,8 @@ ceph::bufferlist Journal::encode_record(
     sizeof(checksum_t),
     reinterpret_cast<const char *>(&metadata_crc_le));
 
+  bl.rebuild();
```
Comment (tchaikov) [inline src/crimson/os/seastore/journal.cc:150]: could you add a comment explaining why we are rebuilding the buffer here?

File: src/crimson/os/seastore/segment_manager/block.cc [hunk 6]
```diff
@@ -46,6 +46,38 @@ static write_ertr::future<> do_write(
...
+  }
+  return device.dma_write(
+    offset,
+    std::move(iov)
+  ).handle_exception([](auto e) -> write_ertr::future<size_t> {
+      logger().error(
+	"do_writev: dma_write got error {}",
+	e);
+      return crimson::ct_error::input_output_error::make();
+  }).then([bl/* hold the buf until the end of io */](auto result)
```
Comment (tchaikov) [inline src/crimson/os/seastore/segment_manager/block.cc:72]:
```suggestion
  }).then([bl/* hold the buf until the end of io */](size_t written)
```

i'd suggest be explicit unless it hurts readability or flexibility.

**Comment by xxhdx1985126** [comment]
@tchaikov Done:-)

File: src/crimson/os/seastore/segment_manager/block.cc (hunk 6 above)
Comment (xxhdx1985126) [reply src/crimson/os/seastore/segment_manager/block.cc]: Done:-)

**Review by tchaikov — APPROVED** [review]

File: src/crimson/os/seastore/segment_manager/block.cc [hunk 7]
```diff
@@ -46,6 +46,38 @@ static write_ertr::future<> do_write(
...
+  logger().debug(
+    "block: do_writev offset {} len {}",
+    offset,
+    bl.length());
+  std::vector<iovec> iov;
+  iov.reserve(bl.get_num_buffers());
+  for (auto& bp : bl.mut_buffers()) {
+    logger().debug("block: do_writev iov length: {}", bp.length());
+    iov.emplace_back(iovec{bp.c_str(), bp.length()});
+  }
```
Comment (cyx1231st) [inline src/crimson/os/seastore/segment_manager/block.cc:63]: nit, we can reuse `bl.prepare_iov(&iov)`;

File: src/crimson/os/seastore/journal.cc [hunk 8]
```diff
@@ -147,6 +147,10 @@ ceph::bufferlist Journal::encode_record(
     sizeof(checksum_t),
     reinterpret_cast<const char *>(&metadata_crc_le));
 
+  // writev requires each buffer to be aligned to the disks' block
+  // size, we need to rebuild here
+  bl.rebuild_aligned(block_size);
```
Comment (cyx1231st) [inline src/crimson/os/seastore/journal.cc:152]: Is there a way to build `bl` in an aligned way, so we don't need to `rebuild_aligned()`?

**Review by cyx1231st — COMMENTED** [review]
Some nits.

File: src/crimson/os/seastore/journal.cc (hunk 8 above)
Comment (xxhdx1985126) [reply src/crimson/os/seastore/journal.cc:152]: Seems that list::reserve() can help, will try

Comment (xxhdx1985126) [reply src/crimson/os/seastore/journal.cc:152]: It seems that `list::reserve()` is not effective in our case, in which we have to encode other bufferlists in the current bl.

Comment (xxhdx1985126) [reply src/crimson/os/seastore/journal.cc:152]: It looks to me that the current bufferlist implementation doesn't support what we are proposing

File: src/crimson/os/seastore/segment_manager/block.cc (hunk 7 above)
Comment (xxhdx1985126) [reply src/crimson/os/seastore/segment_manager/block.cc]: done:-)

**Comment by xxhdx1985126** [comment]
jenkins test make check

**Review by tchaikov — APPROVED** [review]

**Comment by athanatos** [comment]
I don't think write vs writev is a distinction that matters except in BlockSegmentManager::segment_write.  segment_write already allocates a new bufferptr and copies the bl into it, why not just unconditionally call do_writev there?  I don't see any reason to pollute the higher level interfaces with writev.

**Comment by xxhdx1985126** [comment]
> I don't think write vs writev is a distinction that matters except in BlockSegmentManager::segment_write. segment_write already allocates a new bufferptr and copies the bl into it, why not just unconditionally do rebuild_aligned(), do_writev there? I don't see any reason to pollute the higher level interfaces with writev.

Will do.

**Comment by xxhdx1985126** [comment]
> I don't think write vs writev is a distinction that matters except in BlockSegmentManager::segment_write. segment_write already allocates a new bufferptr and copies the bl into it, why not just unconditionally do rebuild_aligned(), do_writev there? I don't see any reason to pollute the higher level interfaces with writev.

done:-)

File: src/crimson/os/seastore/segment_manager/block.cc (hunk 7 above)
Comment (cyx1231st) [reply src/crimson/os/seastore/segment_manager/block.cc:63]: Reverted somehow.

Comment (xxhdx1985126) [reply src/crimson/os/seastore/segment_manager/block.cc:63]: Oh, sorry

**Review by cyx1231st — APPROVED** [review]

File: src/crimson/os/seastore/segment_manager/block.cc [hunk 9]
```diff
@@ -311,29 +339,10 @@ Segment::write_ertr::future<> BlockSegmentManager::segment_write(
...
-  // constituent buffers and they will remain unmodified until the write
-  // completes
-  return seastar::do_with(
-    std::move(bptr),
-    [&](auto &bp) {
-      return do_write(device, get_offset(addr), bp);
-    });
+  // writev requires each buffer to be aligned to the disks' block
+  // size, we need to rebuild here
+  bl.rebuild_aligned(superblock.block_size);
```
[the remark this answers is not in the document]
Comment (xxhdx1985126) [reply src/crimson/os/seastore/segment_manager/block.cc]: done:-)

File: src/crimson/os/seastore/segment_manager/block.cc [hunk 10]
```diff
@@ -46,6 +46,40 @@ static write_ertr::future<> do_write(
...
+  bl.prepare_iov(&iov);
+  return device.dma_write(
+    offset,
+    std::move(iov)
+  ).handle_exception([](auto e) -> write_ertr::future<size_t> {
+      logger().error(
+	"do_writev: dma_write got error {}",
+	e);
+      return crimson::ct_error::input_output_error::make();
+  }).then([bl/* hold the buf until the end of io */](size_t written)
```
Comment (athanatos) [inline src/crimson/os/seastore/segment_manager/block.cc:74]: bl=std::move(bl) would be cheaper.

Comment (xxhdx1985126) [reply src/crimson/os/seastore/segment_manager/block.cc]: done:-)

**Review by athanatos — APPROVED** [review]

**Pull request merged**